Implement more CoAP Options

GiedriusM opened the pull request:
* Observe
* Max-Age
* Uri-Query

GiedriusM commented:
I used these in my application, might be useful for others too.

GiedriusM commented:
Also, I've noticed that you must call option append functions in order by their CoAP code.
This works:
`
otCoapHeaderAppendObserveOption(&reqHeader, 0);
otCoapHeaderAppendUriPathOptions(&reqHeader, "path");
otCoapHeaderAppendMaxAgeOption(&reqHeader, 1234);
otCoapHeaderAppendUriQueryOption(&reqHeader, "key=value");
`

This doesn't:
`
otCoapHeaderAppendUriPathOptions(&reqHeader, "path");
otCoapHeaderAppendUriQueryOption(&reqHeader, "key=value");
otCoapHeaderAppendObserveOption(&reqHeader, 0);
otCoapHeaderAppendMaxAgeOption(&reqHeader, 1234);
`
Observe and Max-Age are not added because kCoapOptionObserve, kCoapOptionMaxAge < kCoapOptionUriQuery.

Should I create a [known-]issue for this?

aeliot commented on src/core/coap/coap_header.cpp:
```diff
@@ -244,6 +244,32 @@ ThreadError Header::AppendOption(const Option &aOption)
+
+    coapOption.mNumber = kCoapOptionObserve;
+
+    if (aObserve <= 0xff) {
```
For `if` statements, the `{` should be on its own line.
```
if (...)
{
    ....
}
else
{
    ....
}
```

aeliot commented on src/core/coap/coap_header.cpp:
```diff
@@ -281,6 +307,49 @@ ThreadError Header::AppendContentFormatOption(MediaType aType)
+
+    coapOption.mNumber = kCoapOptionMaxAge;
+
+    if (aMaxAge <= 0xff) {
```
Same note as above, curly braces should have their own lines.

aeliot commented on src/core/coap/coap_header.cpp:
```diff
@@ -281,6 +307,49 @@ ThreadError Header::AppendContentFormatOption(MediaType aType)
+
+    coapOption.mNumber = kCoapOptionUriQuery;
+    coapOption.mLength = strlen(aUriQuery);
+    coapOption.mValue = (uint8_t *)aUriQuery;
```
Can you align the `=` here?

rlubos commented on src/core/coap/coap_header.cpp:
```diff
@@ -281,6 +312,56 @@ ThreadError Header::AppendContentFormatOption(MediaType aType)
+
+    coapOption.mNumber = kCoapOptionUriQuery;
+    coapOption.mLength = static_cast<uint16_t>(strlen(aUriQuery));
+    coapOption.mValue  = reinterpret_cast<const uint8_t *>(aUriQuery);
```
Is reinterpret_cast really needed here, wouldn't static_cast be enough?

GiedriusM replied:
I was planning on doing static cast, but noticed that AppendUriPathOptions uses reinterpret and just went with common "style".

rlubos replied:
So it must've been missed at some point, as generally static_cast's are preferred. There was even a PR some time ago to get rid of reinterpret_cast wherever possible :)

GiedriusM replied:
Do you want me to replace  reinterpret to static cast in AppendUriPathOptions? Or keep the focus only on the new code?

jwhui replied:
Go ahead and replace with `static_cast()`.  The change is small and contained within the same file.  Thanks!

GiedriusM replied:
Hah, that's the reason for reinterpret_cast:
> error: invalid static_cast from type 'const char*' to type 'const uint8_t* {aka const unsigned char*}'

Looking at http://eel.is/c++draft/basic.types#basic.fundamental probably explains why

> Plain char, signed char, and unsigned char are **three distinct types**, collectively called narrow character types

jwhui replied:
Yes, I remember now.  That's why I had it as `reinterpret_cast` to begin with :)

rlubos commented on src/core/coap/coap_header.cpp:
```diff
@@ -281,6 +312,56 @@ ThreadError Header::AppendContentFormatOption(MediaType aType)
+
+    coapOption.mNumber = kCoapOptionMaxAge;
+
+    if (aMaxAge <= 0xff)
```
Have you considered using a loop to make the code more concise? You could make it like:
```
aMaxAge = Encoding::BigEndian::HostSwap32(aMaxAge);
coapOption.mLength = 0;
do
{
    buf[coapOption.mLength] = aMaxAge & 0xff;
    aMaxAge = aMaxAge >> 8;
    coapOption.mLength++;
} while (aMaxAge != 0 && coapOption.mLength < sizeof(buf))
```
You could even make an generic function from it and use in both, AppendMaxAgeOption and AppendObserveOption.

GiedriusM replied:
Will do

GiedriusM replied:
On a second thought, are you sure this code would work? E.g. given aMaxAge = 0x00123456:
on LE machines: aMaxAge => 0x56341200, buf = { 00, 12, 34, 56 } - good order, length too long
on BE machines: aMaxAge => 0x00123456, buf = { 56, 34, 12 } - reversed order, good length.
Or am I missing something?
I think this should give the correct result:
```C++
aMaxAge = Encoding::BigEndian::HostSwap32(aMaxAge);
coapOption.mLength = 4;
coapOption.mValue = static_cast<uint8_t *>(&aMaxAge);
// skip preceding zeros, but make sure mLength is at least 1
while (coapOption.mValue[0] == 0 && coapOption.mLength > 1) {
    coapOption.mValue++;
    coapOption.mLength--;
}
```
, but not so sure if short and vague is better than long and clear.

GiedriusM replied:
Alright, I'll try to test and push the changes tomorrow.

rlubos replied:
Hey, you're right, I've took it too fast with the example and it had issues you've pointed, sorry about that! I really like the solution you've proposed, especially it seems quite generic and could be reused in both options.

GiedriusM commented on include/openthread-coap.h:
```diff
@@ -224,6 +239,31 @@ ThreadError otCoapHeaderAppendOption(otCoapHeader *aHeader, const otCoapOption *
+ *
+ * @param[inout]  aHeader   A pointer to the CoAP header.
+ * @param[in]     aUriQuery A pointer to NULL-terminated string, which should contain
+ *  a single key=value pair.
```
I didn't want it to be on the same column as @param, so just let the IDE to auto-align it. I'll make it one line, seeing as there are longer lines in that file.

rlubos replied:
By alignment I meant to align it to a column where description begins (some other docs do it this way):
```
* @param[in]     aUriQuery A pointer to NULL-terminated string, which should contain
*                          a single key=value pair.
```
But a single line should work as well 👍

GiedriusM commented:
Sorry I didn't have time to do the changes yesterday, as I said I would. I'm pushing them now. Fixed comment alignment and changed code to be more concise. Unit tested it locally on my machine, seemed to work. Currently don't have access to real setup to sniff if everything is good over-the-air, but I'm pretty sure it will be. Please check if I haven't missed anything.

GiedriusM commented:
@jwhui P.S. this solution saved 38 bytes for Cortex M0+ processor :)

jwhui commented:
@GiedriusM, thanks for the update!

BTWk looks like clang static analyzer is failing on you latest update:
```
coap/coap_header.cpp:186:44: warning: The left operand of '-' is a garbage value
    uint16_t optionDelta = aOption.mNumber - mOptionLast;
```

Also, make sure to run `astyle` before pushing your changes.

GiedriusM commented:
Left an embarrassing mistake. Rushing, while switching between different repos and environments does not help :) Hopefully, it's all good now. Tested in over the air setup.
